File: 2764-maximum-number-of-fish-in-a-grid/maximum-number-of-fish-in-a-grid.java
class Solution {
    int[] dx = {-1, 1, 0, 0};
    int[] dy = {0, 0, 1, -1};
    public int findMaxFish(int[][] g) {
        int n = g.length;
        int m = g[0].length;
        boolean[][] vis = new boolean[n][m] ;
        int ans = 0;
        for(int i = 0 ; i < n ; i ++){
            for(int j = 0 ; j < m ; j ++) if(g[i][j] > 0){
                ans = Math.max(ans, floodfill(i, j, g, vis)) ;
            }
        }
        return ans ;
    }
    private int floodfill(int x, int y, int[][] a, boolean[][] vis){
        if(x < 0 || x >= a.length || y < 0 || y >= a[0].length || a[x][y] == 0 || vis[x][y])
            return 0 ;
        int ans = a[x][y] ;
        vis[x][y] = true ;
        for(int i = 0 ; i < 4 ; i ++){
            ans += floodfill(x + dx[i], y + dy[i], a, vis);
        }
        return ans ;
    }
}

File: 2764-maximum-number-of-fish-in-a-grid/maximum-number-of-fish-in-a-grid.py
class Solution:
    def findMaxFish(self, a: List[List[int]]) -> int:
        n, m = len(a), len(a[0])
        vis = [[False] * m for _ in range(n)]
        dir = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        def floodfill(x, y):
            ans = a[x][y]
            vis[x][y] = True 
            for dx, dy in dir:
                if 0 <= x + dx < n and 0 <= y + dy < m and not vis[x + dx][y + dy] and a[x + dx][y + dy] > 0:
                    ans += floodfill(x + dx, y + dy)
            return ans 

        res = 0 
        for i in range(n):
            for j in range(m):
                if a[i][j] > 0:
                    res = max(floodfill(i, j), res)
        
        return res 